syntax = "proto2";
package msgpb;

import "raft_cmdpb.proto";
import "raft_serverpb.proto";
import "kvrpcpb.proto";
import "coprocessor.proto";
import "volume_cmdpb.proto";
import "pdpb.proto";


enum MessageType {
    Cmd          = 1;
    CmdResp      = 2;
    Raft         = 3;
    KvReq        = 4;
    KvResp       = 5;
    CopReq       = 6;
    CopResp      = 7;
    PdReq        = 8;
    PdResp       = 9;
    VolumeReq       = 10;
    VolumeResp      = 11;
}

// Message holds all messages communicating with TiKV.
message Message {
    optional MessageType                 msg_type  = 1 ;
    optional raft_cmdpb.RaftCmdRequest   cmd_req   = 2;
    optional raft_cmdpb.RaftCmdResponse  cmd_resp  = 3;
    optional raft_serverpb.RaftMessage   raft      = 4;
    optional kvrpcpb.Request             kv_req    = 5;
    optional kvrpcpb.Response            kv_resp   = 6;
    optional coprocessor.Request         cop_req   = 7;
    optional coprocessor.Response        cop_resp  = 8;
    optional pdpb.Request                pd_req    = 9;
    optional pdpb.Response               pd_resp   = 10;
    optional volume_cmdpb.Request        volume_req     = 11;
    optional volume_cmdpb.Response       volume_resp    = 12;
}
